woocommerce/woocommerce-gutenberg-products-block
# Improve inline validation to avoid invalid server pushes

Tagging @opr since he is familiar with the validation system, but others can review it.

When I was working on the error states, I noticed a few problems with the validation system, namely:

- Some client-side validation was not matching up the server side validation
- Invalid data is pushed while you type, resulting in duplicate error messages

To test this, in trunk, go to the checkout and:

1. Make sure you're logged in so the field is populated with a valid email address e.g. `test@test.com`
2. Remove the .com part so it's invalid: `test@test`
3. Wait a few moments and you'll see an error message come back from the server
4. Click out of the field and you'll see no inline validation

We can fix this by allowing custom validation in the client, and pushing only valid states by holding a local state within the text field.

### Testing

#### User Facing Testing

1. Make sure you're logged in so the field is populated with a valid email address e.g. `test@test.com`
2. Remove the .com part so it's invalid: `test@test`
3. Wait a few moments—no errors should appear.
4. Click out of the field. An inline validation error should appear.



### WooCommerce Visibility




### Changelog

> Prevent invalid data being pushed to the server when validating fields on the checkout.
